bikes/views.py

```diff
@@ -82,3 +88,80 @@ def create(self, request, *args, **kwargs):
             {"message": "Bike is not available, it is rented, blocked or reserved"},
             status=status.HTTP_422_UNPROCESSABLE_ENTITY,
         )
+
+
+class ReservationsViewSet(viewsets.ModelViewSet):
+    queryset = Bike.objects.filter(status=BikeStatus.reserved)
+    serializer_class = ReserveBikeSerializer
+
+    def get_serializer_class(self):
+        if self.action == "create":
+            return ReserveIdSerializer
+        else:
+            return ReserveBikeSerializer
+
+    def handle_exception(self, exc):
+        if isinstance(exc, Http404):
+            return Response(
+                {"message": "Bike not reserved"},
+                status=status.HTTP_422_UNPROCESSABLE_ENTITY,
+            )
+
+        return super().handle_exception(exc)
+
+    def create(self, request, *args, **kwargs):
+        serializer = self.get_serializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        try:
+            reserved_bike = Bike.objects.get(id=request.data.get("id"))
+        except Bike.DoesNotExist:
+            return Response(
+                {"message": "Bike not found"}, status=status.HTTP_404_NOT_FOUND
+            )
+        if (
+            reserved_bike.status != BikeStatus.available
+            or reserved_bike.station.state == StationState.blocked
+        ):
+            return Response(
+                {
+                    "message": "Bike already blocked, rented or reserved, or station is blocked"
+                },
+                status=status.HTTP_422_UNPROCESSABLE_ENTITY,
+            )
+        # TODO(kboryczka): uncomment once blocking users is introduced
+        # user = somehow_get_user_from_token_in_headers(request)
+        # if user.status == UserStatus.blocked:
+        #     return Response(
+        #         {"message": "User is blocked"},
+        #         status=status.HTTP_403_FORBIDDEN,
+        #     )
+        time = timezone.now()
+        Reservation.objects.create(
+            bike=reserved_bike,
+            reserved_at=time,
+            reserved_till=time + timezone.timedelta(minutes=30),
+        )
+
+        reserved_bike.status = BikeStatus.reserved
+        reserved_bike.save()
+        return Response(
+            data=ReserveBikeSerializer(reserved_bike).data,
+            status=status.HTTP_201_CREATED,
+        )
+
+    def destroy(self, request, *args, **kwargs):
+        reserved_bike = self.get_object()
+        if reserved_bike.status != BikeStatus.reserved:
+            return Response(
+                {"message": "Bike not reserved"},
+                status=status.HTTP_422_UNPROCESSABLE_ENTITY,
+            )
+        reserved_bike = self.get_object()
+        reserved_bike.reservation.delete()
+
+        reserved_bike.status = BikeStatus.available
+        reserved_bike.save()
+        return Response(
+            data=ReserveBikeSerializer(reserved_bike).data,
+            status=status.HTTP_200_OK,
```

Comment:
Z tego co rozumiem to przy usuwaniu rezerwacji zwracamy wszystkie informacje o rezerwacji (bo tak sprawdzane jest w testach body) tak samo jak przy tworzeniu rezerwacji? W specyfikacji podczas usuwania body jest puste. Poza tym przy poprawnym usunięciu rezerwacji zwracane powinno być 204 według specyfikacji (dzisiaj była aktualizacja endpointów).

Reply:
Tak, Kamil robił z spec v3.1, w v4.0 to się zmieniło.